AMM is incompatible with actors

With
```yaml
distributed:
  scheduler:
    active-memory-manager:
      start: true
      interval: 0.01s
```

Several tests in test_actor.py become very unstable.
For example `test_worker_actions` now fails frequently with

```
>   value = getattr(self.state.actors[actor], attribute)
E   KeyError: 'Counter-73f67bdc-7ad0-43af-9c4a-176532f810ca'
```

--- a/distributed/active_memory_manager.py
+++ b/distributed/active_memory_manager.py
@@ -230,6 +230,10 @@ class ActiveMemoryManagerExtension:
             log_reject(f"ts.state = {ts.state}")
             return None
 
+        if ts.actor:
+            log_reject("task is an actor")
+            return None
+
         if candidates is None:
             candidates = self.scheduler.running.copy()
         else:
@@ -280,6 +284,10 @@ class ActiveMemoryManagerExtension:
 
         if len(ts.who_has) - len(pending_drop) < 2:
             log_reject("less than 2 replicas exist")
+            return None
+
+        if ts.actor:
+            log_reject("task is an actor")
             return None
 
         if candidates is None:
@@ -591,11 +599,26 @@ class RetireWorker(ActiveMemoryManagerPolicy):
             self.manager.policies.remove(self)
             return
 
+        if ws.actors:
+            logger.warning(
+                f"Tried retiring worker {self.address}, but it holds actor(s) "
+                f"{set(ws.actors)}, which can't be moved."
+                "The worker will not be retired."
+            )
+            self.no_recipients = True
+            self.manager.policies.remove(self)
+            return
+
         nrepl = 0
         nno_rec = 0
 
         logger.debug("Retiring %s", ws)
         for ts in ws.has_what:
+            if ts.actor:
+                # This is just a proxy Actor object; if there were any originals we
+                # would have stopped earlier
+                continue
+
             if len(ts.who_has) > 1:
                 # There are already replicas of this key on other workers.
                 # Suggest dropping the replica from this worker.
@@ -663,10 +686,10 @@ class RetireWorker(ActiveMemoryManagerPolicy):
     def done(self) -> bool:
         """Return True if it is safe to close the worker down; False otherwise"""
         if self not in self.manager.policies:
-            # Either the no_recipients flag has been raised, or there were no unique replicas
-            # as of the latest AMM run. Note that due to tasks transitioning from running to
-            # memory there may be some now; it's OK to lose them and just recompute them
-            # somewhere else.
+            # Either the no_recipients flag has been raised, or there were no unique
+            # replicas as of the latest AMM run. Note that due to tasks transitioning
+            # from running to memory there may be some now; it's OK to lose them and
+            # just recompute them somewhere else.
             return True
         ws = self.manager.scheduler.workers.get(self.address)
         if ws is None:

--- a/distributed/deploy/adaptive.py
+++ b/distributed/deploy/adaptive.py
@@ -76,7 +76,7 @@ class Adaptive(AdaptiveCore):
     :meth:`Adaptive.workers_to_close` to control when the cluster should be
     resized. The default implementation checks if there are too many tasks
     per worker or too little memory available (see
-    :meth:`Scheduler.adaptive_target`).
+    :meth:`distributed.Scheduler.adaptive_target`).
     The values for interval, min, max, wait_count and target_duration can be
     specified in the dask config under the distributed.adaptive key.
     '''

--- a/distributed/scheduler.py
+++ b/distributed/scheduler.py
@@ -1108,7 +1108,7 @@ class TaskState:
     #:     "processing" state and be sent for execution to another connected worker.
     loose_restrictions: bool
 
-    #: Whether or not this task is an Actor
+    #: Whether this task is an Actor
     actor: bool
 
     #: The group of tasks to which this one belongs

--- a/distributed/tests/test_active_memory_manager.py
+++ b/distributed/tests/test_active_memory_manager.py
@@ -1056,6 +1056,78 @@ async def test_RetireWorker_faulty_recipient(c, s, *nannies):
     clog_fut.cancel()
 
 
+class Counter:
+    def __init__(self):
+        self.n = 0
+
+    def increment(self):
+        self.n += 1
+
+
+@gen_cluster(client=True, config=demo_config("drop"))
+async def test_dont_drop_actors(c, s, a, b):
+    x = c.submit(Counter, key="x", actor=True, workers=[a.address])
+    y = c.submit(lambda cnt: cnt.increment(), x, key="y", workers=[b.address])
+    await wait([x, y])
+    assert len(s.tasks["x"].who_has) == 2
+    s.extensions["amm"].run_once()
+    await asyncio.sleep(0.2)
+    assert len(s.tasks["x"].who_has) == 2
+
+
+@gen_cluster(client=True, config=demo_config("replicate"))
+async def test_dont_replicate_actors(c, s, a, b):
+    x = c.submit(Counter, key="x", actor=True)
+    await wait(x)
+    assert len(s.tasks["x"].who_has) == 1
+    s.extensions["amm"].run_once()
+    await asyncio.sleep(0.2)
+    assert len(s.tasks["x"].who_has) == 1
+
+
+@pytest.mark.parametrize("has_proxy", [False, True])
+@gen_cluster(client=True, config=NO_AMM_START)
+async def test_RetireWorker_with_actor(c, s, a, b, has_proxy):
+    """A worker holding one or more original actor objects cannot be retired"""
+    x = c.submit(Counter, key="x", actor=True, workers=[a.address])
+    await wait(x)
+    assert "x" in a.state.actors
+
+    if has_proxy:
+        y = c.submit(
+            lambda cnt: cnt.increment().result(), x, key="y", workers=[b.address]
+        )
+        await wait(y)
+        assert "x" in b.data
+        assert "y" in b.data
+
+    with captured_logger("distributed.active_memory_manager", logging.WARNING) as log:
+        out = await c.retire_workers([a.address])
+    assert out == {}
+    assert "it holds actor(s)" in log.getvalue()
+    assert "x" in a.state.actors
+
+    if has_proxy:
+        assert "x" in b.data
+        assert "y" in b.data
+
+
+@gen_cluster(client=True, config=NO_AMM_START)
+async def test_RetireWorker_with_actor_proxy(c, s, a, b):
+    """A worker holding an Actor proxy object can be retired as normal."""
+    x = c.submit(Counter, key="x", actor=True, workers=[a.address])
+    y = c.submit(lambda cnt: cnt.increment().result(), x, key="y", workers=[b.address])
+    await wait(y)
+    assert "x" in a.state.actors
+    assert "x" in b.data
+    assert "y" in b.data
+
+    out = await c.retire_workers([b.address])
+    assert out.keys() == {b.address}
+    assert "x" in a.state.actors
+    assert "y" in a.data
+
+
 class DropEverything(ActiveMemoryManagerPolicy):
     """Inanely suggest to drop every single key in the cluster"""
 

--- a/distributed/worker_state_machine.py
+++ b/distributed/worker_state_machine.py
@@ -2892,7 +2892,7 @@ class WorkerState:
     @_handle_event.register
     def _handle_gather_dep_success(self, ev: GatherDepSuccessEvent) -> RecsInstrs:
         """gather_dep terminated successfully.
-        The response may contain less keys than the request.
+        The response may contain fewer keys than the request.
         """
         recommendations: Recs = {}
         for ts in self._gather_dep_done_common(ev):

--- a/docs/source/actors.rst
+++ b/docs/source/actors.rst
@@ -227,3 +227,7 @@ Actors offer advanced capabilities, but with some cost:
     computations no diagnostics are available about these computations.
 3.  **No Load balancing:** Actors are allocated onto workers evenly, without
     serious consideration given to avoiding communication.
+4.  **No dynamic clusters:** Actors cannot be migrated to other workers.
+    A worker holding an actor can be retired neither through
+    :meth:`~distributed.Client.retire_workers` nor through
+    :class:`~distributed.deploy.Adaptive`.

